frontend/src/editor/commands/addBorderNumbers.ts

```diff
@@ -17,114 +12,55 @@ function addBorderNumbers({ state, dispatch }: CommandProps): boolean {
   const schema = state.schema
 
   const paragraphNodeType = schema.nodes.paragraph
+  const numberNodeType = schema.nodes.borderNumberNumber
   const contentNodeType = schema.nodes.borderNumberContent
+  const borderNumberNodeType = schema.nodes.borderNumber
 
   let modified = false
-  let shift = 0
-  let updatedFrom = initialFrom
 
-  state.doc.nodesBetween(initialFrom, initialTo, (node, pos) => {
-    if (hasBorderNumber(state, pos, schema)) {
-      return
-    }
+  const nodesWithoutBorderNumbers: { node: ProsemirrorNode; pos: number }[] = []
+  state.doc.nodesBetween(initialFrom, initialTo, (node, pos, parent) => {
+    const isTopLevelNode = parent && parent.type === state.schema.topNodeType
 
-    if (isParagraphWithContent(node, paragraphNodeType)) {
-      const borderNumberNode = createBorderNumberNode(
-        node,
-        schema,
-        contentNodeType,
-      )
-      const currentPos = pos + shift
+    const isBorderNumberNode =
+      node.type === borderNumberNodeType || node.hasMarkup(borderNumberNodeType)
```

Comment:
@kiwikern I don't understand why we need node.hasMarkup here?

Reply:
Does not make sense, can be removed. 👍

Reply:
Done